Enphase Envoy: Configurable scan interval

boc-the-git opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This allows a user to set the scan interval for their Enphase Envoy system. With the recent introductions of the Energy integration, the 60 seconds hardcoded interval was a hindrance to data accuracy.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

boc-the-git committed:
Introduce a configurable scan interval

boc-the-git committed:
Need to commit the constants..

boc-the-git committed:
Unit test

boc-the-git committed:
Fix async_unload

boc-the-git committed:
Merge branch 'home-assistant:dev' into configurable_scan_interval

boc-the-git committed:
Linting fixes

boc-the-git committed:
Merge branch 'home-assistant:dev' into configurable_scan_interval

boc-the-git commented:
Need a bit longer here than originally planned.. changes in 2021.9 causing me some hiccups

boc-the-git commented:
The issue I'm experiencing on my test server seems to be this (https://github.com/home-assistant/core/issues/50872).. it was resolved by modifying .storage/core.config_entries
As such I think I'll proceed with tidying this change, as I initially feared I was causing that issue.

ludeeus reviewed (changes requested):
We should not implement this, there is a way for the user to disable polling and implement their own schedule if they want, <https://www.home-assistant.io/blog/2021/06/02/release-20216/#disable-polling-updates-on-any-integration>

boc-the-git committed:
Fix strings

boc-the-git commented:
> We should not implement this, there is a way for the user to disable polling and implement their own schedule if they want, https://www.home-assistant.io/blog/2021/06/02/release-20216/#disable-polling-updates-on-any-integration

Well, that's certainly something I need to look into more!
Initial impressions are that it might not work for the enphase 'device', but will dig into it more in the coming days.

Thanks

boc-the-git commented:
@ludeeus I can't see anything that would limit update_entity to every a minimum of every 10seconds, but that seems to be happening. Is there something I'm missing?

With my change here, I have updates every 2s:
`2021-09-04 10:29:10 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Finished fetching envoy Envoy data in 0.360 seconds (success: True)
2021-09-04 10:29:12 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Finished fetching envoy Envoy data in 0.316 seconds (success: True)
2021-09-04 10:29:14 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Finished fetching envoy Envoy data in 0.620 seconds (success: True)
2021-09-04 10:29:16 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Finished fetching envoy Envoy data in 0.345 seconds (success: True)`

This is when using an automation triggering every 2s, you can see all 3 logs 'call the service' (Event call_service) but only at 09:35:08 does it actually get through to enphase:
`2021-09-04 09:35:06 DEBUG (MainThread) [homeassistant.components.automation.enphase_2secs] Automation triggered by time pattern
2021-09-04 09:35:06 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event automation_triggered[L]: name=enphase 2secs, entity_id=automation.enphase_2secs, source=time pattern>
2021-09-04 09:35:06 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=automation.enphase_2secs, old_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:04.005697+10:00, mode=single, current=0, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>, new_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:06.003067+10:00, mode=single, current=1, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>>
2021-09-04 09:35:06 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 09:35:06 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 09:35:06 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "automation.enphase_2secs", "old_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:04.005697+00:00", "mode": "single", "current": 0, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:04.011674+00:00", "context": {"id": "6fd698a04bf3d88d2b629a37b9c0e90c", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:06.003067+00:00", "mode": "single", "current": 1, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:06.003199+00:00", "context": {"id": "a94ad2770a6562a474d671cc3ac33f22", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:06.003199+00:00", "context": {"id": "a94ad2770a6562a474d671cc3ac33f22", "parent_id": null, "user_id": null}}}
2021-09-04 09:35:06 DEBUG (Recorder) [homeassistant.components.recorder] Sending keepalive
2021-09-04 09:35:06 DEBUG (Recorder) [homeassistant.components.recorder.pool.RecorderPool] Connection <sqlite3.Connection object at 0x7fe7d3b448a0> checked out from pool
2021-09-04 09:35:06 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=homeassistant, service=update_entity, service_data=entity_id=['sensor.envoy_current_power_production', 'sensor.envoy_today_s_energy_production', 'sensor.envoy_last_seven_days_energy_production', 'sensor.envoy_lifetime_energy_production', 'sensor.envoy_current_power_consumption', 'sensor.envoy_today_s_energy_consumption', 'sensor.envoy_last_seven_days_energy_consumption', 'sensor.envoy_lifetime_energy_consumption']>
2021-09-04 09:35:06 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=automation.enphase_2secs, old_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:06.003067+10:00, mode=single, current=1, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>, new_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:06.003067+10:00, mode=single, current=0, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>>
2021-09-04 09:35:06 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "automation.enphase_2secs", "old_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:06.003067+00:00", "mode": "single", "current": 1, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:06.003199+00:00", "context": {"id": "a94ad2770a6562a474d671cc3ac33f22", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:06.003067+00:00", "mode": "single", "current": 0, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:06.016351+00:00", "context": {"id": "a94ad2770a6562a474d671cc3ac33f22", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:06.016351+00:00", "context": {"id": "a94ad2770a6562a474d671cc3ac33f22", "parent_id": null, "user_id": null}}}
<..deleted a number of zeroconf related lines..>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.automation.enphase_2secs] Automation triggered by time pattern
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event automation_triggered[L]: name=enphase 2secs, entity_id=automation.enphase_2secs, source=time pattern>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=automation.enphase_2secs, old_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:06.003067+10:00, mode=single, current=0, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>, new_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:08.003508+10:00, mode=single, current=1, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>>
2021-09-04 09:35:08 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 09:35:08 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "automation.enphase_2secs", "old_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:06.003067+00:00", "mode": "single", "current": 0, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:06.016351+00:00", "context": {"id": "a94ad2770a6562a474d671cc3ac33f22", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:08.003508+00:00", "mode": "single", "current": 1, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:08.003642+00:00", "context": {"id": "13ac229771ac6c9cf85b8b2aed8546c1", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:08.003642+00:00", "context": {"id": "13ac229771ac6c9cf85b8b2aed8546c1", "parent_id": null, "user_id": null}}}
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=homeassistant, service=update_entity, service_data=entity_id=['sensor.envoy_current_power_production', 'sensor.envoy_today_s_energy_production', 'sensor.envoy_last_seven_days_energy_production', 'sensor.envoy_lifetime_energy_production', 'sensor.envoy_current_power_consumption', 'sensor.envoy_today_s_energy_consumption', 'sensor.envoy_last_seven_days_energy_consumption', 'sensor.envoy_lifetime_energy_consumption']>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=automation.enphase_2secs, old_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:08.003508+10:00, mode=single, current=1, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>, new_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:08.003508+10:00, mode=single, current=0, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "automation.enphase_2secs", "old_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:08.003508+00:00", "mode": "single", "current": 1, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:08.003642+00:00", "context": {"id": "13ac229771ac6c9cf85b8b2aed8546c1", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:08.003508+00:00", "mode": "single", "current": 0, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:08.019376+00:00", "context": {"id": "13ac229771ac6c9cf85b8b2aed8546c1", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:08.019376+00:00", "context": {"id": "13ac229771ac6c9cf85b8b2aed8546c1", "parent_id": null, "user_id": null}}}
2021-09-04 09:35:08 DEBUG (MainThread) [httpx._client] HTTP Request: GET http://192.168.2.220/production.json "HTTP/1.1 200 OK"
2021-09-04 09:35:08 DEBUG (MainThread) [envoy_reader.envoy_reader] Fetched from http://192.168.2.220/production.json: <Response [200 OK]>: {redacted}
2021-09-04 09:35:08 DEBUG (MainThread) [httpx._client] HTTP Request: GET http://192.168.2.220/api/v1/production/inverters "HTTP/1.1 401 Unauthorized"
2021-09-04 09:35:08 DEBUG (MainThread) [httpx._client] HTTP Request: GET http://192.168.2.220/api/v1/production/inverters "HTTP/1.1 200 OK"
2021-09-04 09:35:08 DEBUG (MainThread) [envoy_reader.envoy_reader] Fetched from http://192.168.2.220/api/v1/production/inverters: <Response [200 OK]>: [
  redacted
]

2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Retrieved data from API: {'production': 1850, 'daily_production': 2044, 'seven_days_production': redacted, 'lifetime_production': redacted, 'consumption': 135, 'daily_consumption': 2030, 'seven_days_consumption': redacted, 'lifetime_consumption': redacted, 'inverters_production': {redacted}}
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Finished fetching envoy Envoy data in 0.398 seconds (success: True)
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=sensor.envoy_current_power_production, old_state=<state sensor.envoy_current_power_production=1818; state_class=measurement, unit_of_measurement=W, friendly_name=Envoy Current Power Production, icon=mdi:flash @ 2021-09-04T09:34:57.830760+10:00>, new_state=<state sensor.envoy_current_power_production=1850; state_class=measurement, unit_of_measurement=W, friendly_name=Envoy Current Power Production, icon=mdi:flash @ 2021-09-04T09:35:08.234924+10:00>>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=sensor.envoy_today_s_energy_production, old_state=<state sensor.envoy_today_s_energy_production=2039; state_class=measurement, unit_of_measurement=Wh, friendly_name=Envoy Today's Energy Production, icon=mdi:flash, device_class=energy @ 2021-09-04T09:34:57.831559+10:00>, new_state=<state sensor.envoy_today_s_energy_production=2044; state_class=measurement, unit_of_measurement=Wh, friendly_name=Envoy Today's Energy Production, icon=mdi:flash, device_class=energy @ 2021-09-04T09:35:08.235564+10:00>>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=sensor.envoy_last_seven_days_energy_production, old_state=<state sensor.envoy_last_seven_days_energy_production=redacted; state_class=measurement, unit_of_measurement=Wh, friendly_name=Envoy Last Seven Days Energy Production, icon=mdi:flash, device_class=energy @ 2021-09-04T09:34:57.831986+10:00>, new_state=<state sensor.envoy_last_seven_days_energy_production=redacted; state_class=measurement, unit_of_measurement=Wh, friendly_name=Envoy Last Seven Days Energy Production, icon=mdi:flash, device_class=energy @ 2021-09-04T09:35:08.236290+10:00>>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=sensor.envoy_lifetime_energy_production, old_state=<state sensor.envoy_lifetime_energy_production=redacted; state_class=total_increasing, unit_of_measurement=Wh, friendly_name=Envoy Lifetime Energy Production, icon=mdi:flash, device_class=energy @ 2021-09-04T09:34:57.832271+10:00>, new_state=<state sensor.envoy_lifetime_energy_production=redacted; state_class=total_increasing, unit_of_measurement=Wh, friendly_name=Envoy Lifetime Energy Production, icon=mdi:flash, device_class=energy @ 2021-09-04T09:35:08.238407+10:00>>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=sensor.envoy_current_power_consumption, old_state=<state sensor.envoy_current_power_consumption=133; state_class=measurement, unit_of_measurement=W, friendly_name=Envoy Current Power Consumption, icon=mdi:flash @ 2021-09-04T09:34:57.832621+10:00>, new_state=<state sensor.envoy_current_power_consumption=135; state_class=measurement, unit_of_measurement=W, friendly_name=Envoy Current Power Consumption, icon=mdi:flash @ 2021-09-04T09:35:08.239162+10:00>>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=sensor.envoy_today_s_energy_consumption, old_state=<state sensor.envoy_today_s_energy_consumption=2029; state_class=measurement, unit_of_measurement=Wh, friendly_name=Envoy Today's Energy Consumption, icon=mdi:flash, device_class=energy @ 2021-09-04T09:34:47.473567+10:00>, new_state=<state sensor.envoy_today_s_energy_consumption=2030; state_class=measurement, unit_of_measurement=Wh, friendly_name=Envoy Today's Energy Consumption, icon=mdi:flash, device_class=energy @ 2021-09-04T09:35:08.239691+10:00>>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=sensor.envoy_last_seven_days_energy_consumption, old_state=<state sensor.envoy_last_seven_days_energy_consumption=redacted; state_class=measurement, unit_of_measurement=Wh, friendly_name=Envoy Last Seven Days Energy Consumption, icon=mdi:flash, device_class=energy @ 2021-09-04T09:34:47.473978+10:00>, new_state=<state sensor.envoy_last_seven_days_energy_consumption=redacted; state_class=measurement, unit_of_measurement=Wh, friendly_name=Envoy Last Seven Days Energy Consumption, icon=mdi:flash, device_class=energy @ 2021-09-04T09:35:08.239994+10:00>>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=sensor.envoy_lifetime_energy_consumption, old_state=<state sensor.envoy_lifetime_energy_consumption=redacted; state_class=total_increasing, unit_of_measurement=Wh, friendly_name=Envoy Lifetime Energy Consumption, icon=mdi:flash, device_class=energy @ 2021-09-04T09:34:47.474436+10:00>, new_state=<state sensor.envoy_lifetime_energy_consumption=redacted; state_class=total_increasing, unit_of_measurement=Wh, friendly_name=Envoy Lifetime Energy Consumption, icon=mdi:flash, device_class=energy @ 2021-09-04T09:35:08.240312+10:00>>
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "sensor.envoy_current_power_production", "old_state": {"entity_id": "sensor.envoy_current_power_production", "state": "1818", "attributes": {"state_class": "measurement", "unit_of_measurement": "W", "friendly_name": "Envoy Current Power Production", "icon": "mdi:flash"}, "last_changed": "2021-09-03T23:34:57.830760+00:00", "last_updated": "2021-09-03T23:34:57.830760+00:00", "context": {"id": "7ea782acc0122d5453589089736bb215", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "sensor.envoy_current_power_production", "state": "1850", "attributes": {"state_class": "measurement", "unit_of_measurement": "W", "friendly_name": "Envoy Current Power Production", "icon": "mdi:flash"}, "last_changed": "2021-09-03T23:35:08.234924+00:00", "last_updated": "2021-09-03T23:35:08.234924+00:00", "context": {"id": "07689c3321acb1e39316ea07f141bc94", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:08.234924+00:00", "context": {"id": "07689c3321acb1e39316ea07f141bc94", "parent_id": null, "user_id": null}}}
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "sensor.envoy_today_s_energy_production", "old_state": {"entity_id": "sensor.envoy_today_s_energy_production", "state": "2039", "attributes": {"state_class": "measurement", "unit_of_measurement": "Wh", "friendly_name": "Envoy Today's Energy Production", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:34:57.831559+00:00", "last_updated": "2021-09-03T23:34:57.831559+00:00", "context": {"id": "330f960c422f0c394838a08265da18bf", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "sensor.envoy_today_s_energy_production", "state": "2044", "attributes": {"state_class": "measurement", "unit_of_measurement": "Wh", "friendly_name": "Envoy Today's Energy Production", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:35:08.235564+00:00", "last_updated": "2021-09-03T23:35:08.235564+00:00", "context": {"id": "d72f0421e4dc763fcc92677df2078509", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:08.235564+00:00", "context": {"id": "d72f0421e4dc763fcc92677df2078509", "parent_id": null, "user_id": null}}}
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "sensor.envoy_last_seven_days_energy_production", "old_state": {"entity_id": "sensor.envoy_last_seven_days_energy_production", "state": "redacted", "attributes": {"state_class": "measurement", "unit_of_measurement": "Wh", "friendly_name": "Envoy Last Seven Days Energy Production", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:34:57.831986+00:00", "last_updated": "2021-09-03T23:34:57.831986+00:00", "context": {"id": "6feb9752137e951b8d51c239037ee05e", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "sensor.envoy_last_seven_days_energy_production", "state": "redacted", "attributes": {"state_class": "measurement", "unit_of_measurement": "Wh", "friendly_name": "Envoy Last Seven Days Energy Production", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:35:08.236290+00:00", "last_updated": "2021-09-03T23:35:08.236290+00:00", "context": {"id": "d227981cfc05f6c16b20b1b37d3e61ef", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:08.236290+00:00", "context": {"id": "d227981cfc05f6c16b20b1b37d3e61ef", "parent_id": null, "user_id": null}}}
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "sensor.envoy_lifetime_energy_production", "old_state": {"entity_id": "sensor.envoy_lifetime_energy_production", "state": "redacted", "attributes": {"state_class": "total_increasing", "unit_of_measurement": "Wh", "friendly_name": "Envoy Lifetime Energy Production", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:34:57.832271+00:00", "last_updated": "2021-09-03T23:34:57.832271+00:00", "context": {"id": "c3fdde148937b890bdaa5399570325b4", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "sensor.envoy_lifetime_energy_production", "state": "redacted", "attributes": {"state_class": "total_increasing", "unit_of_measurement": "Wh", "friendly_name": "Envoy Lifetime Energy Production", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:35:08.238407+00:00", "last_updated": "2021-09-03T23:35:08.238407+00:00", "context": {"id": "2a1737413297feda8972e8cd91597ab7", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:08.238407+00:00", "context": {"id": "2a1737413297feda8972e8cd91597ab7", "parent_id": null, "user_id": null}}}
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "sensor.envoy_current_power_consumption", "old_state": {"entity_id": "sensor.envoy_current_power_consumption", "state": "133", "attributes": {"state_class": "measurement", "unit_of_measurement": "W", "friendly_name": "Envoy Current Power Consumption", "icon": "mdi:flash"}, "last_changed": "2021-09-03T23:34:57.832621+00:00", "last_updated": "2021-09-03T23:34:57.832621+00:00", "context": {"id": "33ce9bf41584209a41cbce011df946d4", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "sensor.envoy_current_power_consumption", "state": "135", "attributes": {"state_class": "measurement", "unit_of_measurement": "W", "friendly_name": "Envoy Current Power Consumption", "icon": "mdi:flash"}, "last_changed": "2021-09-03T23:35:08.239162+00:00", "last_updated": "2021-09-03T23:35:08.239162+00:00", "context": {"id": "dc0a4ab2fd42e74d4c8ede00e96e04d8", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:08.239162+00:00", "context": {"id": "dc0a4ab2fd42e74d4c8ede00e96e04d8", "parent_id": null, "user_id": null}}}
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "sensor.envoy_today_s_energy_consumption", "old_state": {"entity_id": "sensor.envoy_today_s_energy_consumption", "state": "2029", "attributes": {"state_class": "measurement", "unit_of_measurement": "Wh", "friendly_name": "Envoy Today's Energy Consumption", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:34:47.473567+00:00", "last_updated": "2021-09-03T23:34:47.473567+00:00", "context": {"id": "2010bdb6e8027243cf391b2c0649d267", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "sensor.envoy_today_s_energy_consumption", "state": "2030", "attributes": {"state_class": "measurement", "unit_of_measurement": "Wh", "friendly_name": "Envoy Today's Energy Consumption", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:35:08.239691+00:00", "last_updated": "2021-09-03T23:35:08.239691+00:00", "context": {"id": "148d3451e11550d8e5be01a76508ae7d", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:08.239691+00:00", "context": {"id": "148d3451e11550d8e5be01a76508ae7d", "parent_id": null, "user_id": null}}}
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "sensor.envoy_last_seven_days_energy_consumption", "old_state": {"entity_id": "sensor.envoy_last_seven_days_energy_consumption", "state": "redacted", "attributes": {"state_class": "measurement", "unit_of_measurement": "Wh", "friendly_name": "Envoy Last Seven Days Energy Consumption", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:34:47.473978+00:00", "last_updated": "2021-09-03T23:34:47.473978+00:00", "context": {"id": "f377441990ea609aa103fbd63c60a8b0", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "sensor.envoy_last_seven_days_energy_consumption", "state": "redacted", "attributes": {"state_class": "measurement", "unit_of_measurement": "Wh", "friendly_name": "Envoy Last Seven Days Energy Consumption", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:35:08.239994+00:00", "last_updated": "2021-09-03T23:35:08.239994+00:00", "context": {"id": "17b235d19d480ebb593ac9ee67da3f43", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:08.239994+00:00", "context": {"id": "17b235d19d480ebb593ac9ee67da3f43", "parent_id": null, "user_id": null}}}
2021-09-04 09:35:08 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "sensor.envoy_lifetime_energy_consumption", "old_state": {"entity_id": "sensor.envoy_lifetime_energy_consumption", "state": "redacted", "attributes": {"state_class": "total_increasing", "unit_of_measurement": "Wh", "friendly_name": "Envoy Lifetime Energy Consumption", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:34:47.474436+00:00", "last_updated": "2021-09-03T23:34:47.474436+00:00", "context": {"id": "de6c493c1dd1fd9c0950badb1e9a189e", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "sensor.envoy_lifetime_energy_consumption", "state": "redacted", "attributes": {"state_class": "total_increasing", "unit_of_measurement": "Wh", "friendly_name": "Envoy Lifetime Energy Consumption", "icon": "mdi:flash", "device_class": "energy"}, "last_changed": "2021-09-03T23:35:08.240312+00:00", "last_updated": "2021-09-03T23:35:08.240312+00:00", "context": {"id": "cff70ab19fc00c32bdd0afd62f5e5307", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:08.240312+00:00", "context": {"id": "cff70ab19fc00c32bdd0afd62f5e5307", "parent_id": null, "user_id": null}}}
<..deleted a bunch of zeroconf lines..>
2021-09-04 09:35:10 DEBUG (MainThread) [homeassistant.components.automation.enphase_2secs] Automation triggered by time pattern
2021-09-04 09:35:10 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event automation_triggered[L]: name=enphase 2secs, entity_id=automation.enphase_2secs, source=time pattern>
2021-09-04 09:35:10 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=automation.enphase_2secs, old_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:08.003508+10:00, mode=single, current=0, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>, new_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:10.002898+10:00, mode=single, current=1, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>>
2021-09-04 09:35:10 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 09:35:10 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 09:35:10 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "automation.enphase_2secs", "old_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:08.003508+00:00", "mode": "single", "current": 0, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:08.019376+00:00", "context": {"id": "13ac229771ac6c9cf85b8b2aed8546c1", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:10.002898+00:00", "mode": "single", "current": 1, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:10.003031+00:00", "context": {"id": "fdaa093a4b9df380098923bfbb93a3ef", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:10.003031+00:00", "context": {"id": "fdaa093a4b9df380098923bfbb93a3ef", "parent_id": null, "user_id": null}}}
2021-09-04 09:35:10 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event call_service[L]: domain=homeassistant, service=update_entity, service_data=entity_id=['sensor.envoy_current_power_production', 'sensor.envoy_today_s_energy_production', 'sensor.envoy_last_seven_days_energy_production', 'sensor.envoy_lifetime_energy_production', 'sensor.envoy_current_power_consumption', 'sensor.envoy_today_s_energy_consumption', 'sensor.envoy_last_seven_days_energy_consumption', 'sensor.envoy_lifetime_energy_consumption']>
2021-09-04 09:35:10 DEBUG (MainThread) [homeassistant.core] Bus:Handling <Event state_changed[L]: entity_id=automation.enphase_2secs, old_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:10.002898+10:00, mode=single, current=1, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>, new_state=<state automation.enphase_2secs=on; last_triggered=2021-09-04T09:35:10.002898+10:00, mode=single, current=0, id=1630711190280, friendly_name=enphase 2secs @ 2021-09-04T09:34:36.026465+10:00>>
2021-09-04 09:35:10 DEBUG (MainThread) [homeassistant.components.websocket_api.http.connection] [140633403260352] Sending {"id": 2, "type": "event", "event": {"event_type": "state_changed", "data": {"entity_id": "automation.enphase_2secs", "old_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:10.002898+00:00", "mode": "single", "current": 1, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:10.003031+00:00", "context": {"id": "fdaa093a4b9df380098923bfbb93a3ef", "parent_id": null, "user_id": null}}, "new_state": {"entity_id": "automation.enphase_2secs", "state": "on", "attributes": {"last_triggered": "2021-09-03T23:35:10.002898+00:00", "mode": "single", "current": 0, "id": "1630711190280", "friendly_name": "enphase 2secs"}, "last_changed": "2021-09-03T23:34:36.026465+00:00", "last_updated": "2021-09-03T23:35:10.019470+00:00", "context": {"id": "fdaa093a4b9df380098923bfbb93a3ef", "parent_id": null, "user_id": null}}}, "origin": "LOCAL", "time_fired": "2021-09-03T23:35:10.019470+00:00", "context": {"id": "fdaa093a4b9df380098923bfbb93a3ef", "parent_id": null, "user_id": null}}}
<..deleted a bunch of zeroconf lines..>`

boc-the-git commented:
Also a less verbose version of the automation logging:
```2021-09-04 10:54:56 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:54:56 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:54:56 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Retrieved data from API: {redacted}
2021-09-04 10:54:56 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Finished fetching envoy Envoy data in 0.322 seconds (success: True)
2021-09-04 10:54:58 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:54:58 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:55:00 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:55:00 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:55:02 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:55:02 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:55:04 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:55:04 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:55:06 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:55:06 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:55:06 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Retrieved data from API: {redacted}
2021-09-04 10:55:06 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Finished fetching envoy Envoy data in 0.368 seconds (success: True)
2021-09-04 10:55:08 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:55:08 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:55:10 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:55:10 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:55:12 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:55:12 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:55:14 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:55:14 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:55:16 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:55:16 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:55:17 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Retrieved data from API: {redacted}
2021-09-04 10:55:17 DEBUG (MainThread) [homeassistant.components.enphase_envoy] Finished fetching envoy Envoy data in 0.541 seconds (success: True)
2021-09-04 10:55:18 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:55:18 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
2021-09-04 10:55:20 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Running automation actions
2021-09-04 10:55:20 INFO (MainThread) [homeassistant.components.automation.enphase_2secs] enphase 2secs: Executing step call service
```

MartinHjelmare commented:
The coordinator has a default debounce time of 10 seconds when requesting a refresh:
https://github.com/home-assistant/core/blob/0dc8fb497b1162c08286dfd03f54da3d237f38dd/homeassistant/helpers/update_coordinator.py#L23

boc-the-git commented:
> The coordinator has a default debounce time of 10 seconds when requesting a refresh:
> https://github.com/home-assistant/core/blob/0dc8fb497b1162c08286dfd03f54da3d237f38dd/homeassistant/helpers/update_coordinator.py#L23

Thanks Martin. Is the suggestion then that the enphase component should specify a debouncer with a different cooldown, as a way around this? https://github.com/home-assistant/core/blob/dev/homeassistant/components/enphase_envoy/__init__.py#L64-L70
If that seems a viable path forward, I can look to implement and test it.

MartinHjelmare commented:
Possibly, but we don't allow faster polling than 5 seconds.

Why does this integration need to poll faster than 10 seconds?

boc-the-git commented:
It is reading from a data source that updates rapidly (a solar inverter). The inverter itself does store daily values for some things, so they don't require frequent updates.

The only way I've found to calculate grid consumption/return is to have a template sensor doing the maths based on 'consumption' and 'production' - if we update those values every 10 seconds, we lose 9 seconds of granular data. (and the current 60 seconds from the component is very inadequate)
